main/test/ca/uwaterloo/flix/library/TestDelayMap.flix

```diff
@@ -112,4 +112,471 @@ namespace TestDelayMap {
+
+    @test
+    def unionWith05(): Bool =
+        DelayMap.unionWith((v1, v2) -> v1 - v2, Map.toDelayMap(Map#{1 -> 2, 3 -> 4}), Map.toDelayMap(Map#{})) |> DelayMap.toMap == Map#{1 -> 2, 3 -> 4}
+
+    @test
+    def unionWith06(): Bool =
+        DelayMap.unionWith((v1, v2) -> v1 - v2, Map.toDelayMap(Map#{1 -> 2}), Map.toDelayMap(Map#{3 -> 4})) |> DelayMap.toMap == Map#{1 -> 2, 3 -> 4}
+
+    @test
+    def unionWith07(): Bool =
+        DelayMap.unionWith((v1, v2) -> v1 - v2, Map.toDelayMap(Map#{1 -> 2}), Map.toDelayMap(Map#{1 -> 5})) |> DelayMap.toMap == Map#{1 -> -3}
+
+    @test
+    def unionWith08(): Bool =
+        DelayMap.unionWith((v1, v2) -> v1 - v2, Map.toDelayMap(Map#{}), Map.toDelayMap(Map#{1 -> 2, 2 -> 3, 3 -> 4})) |> DelayMap.toMap == Map#{1 -> 2, 2 -> 3, 3 -> 4}
+
+    @test
+    def unionWith09(): Bool =
+        DelayMap.unionWith((v1, v2) -> v1 - v2, Map.toDelayMap(Map#{1 -> 2, 2 -> 3, 3 -> 4}), Map.toDelayMap(Map#{})) |> DelayMap.toMap == Map#{1 -> 2, 2 -> 3, 3 -> 4}
+
+    @test
+    def unionWith10(): Bool =
+        DelayMap.unionWith((v1, v2) -> v1 - v2, Map.toDelayMap(Map#{1 -> 2, 2 -> 3}), Map.toDelayMap(Map#{3 -> 4})) |> DelayMap.toMap == Map#{1 -> 2, 2 -> 3, 3 -> 4}
+
+    @test
+    def unionWith11(): Bool =
+        DelayMap.unionWith((v1, v2) -> v1 - v2, Map.toDelayMap(Map#{3 -> 4}), Map.toDelayMap(Map#{1 -> 2, 2 -> 3})) |> DelayMap.toMap == Map#{1 -> 2, 2 -> 3, 3 -> 4}
+
+    @test
+    def unionWith12(): Bool =
+        DelayMap.unionWith((v1, v2) -> v1 - v2, Map.toDelayMap(Map#{1 -> 2, 2 -> 3}), Map.toDelayMap(Map#{2 -> 8})) |> DelayMap.toMap == Map#{1 -> 2, 2 -> -5}
+
+    @test
+    def unionWith13(): Bool =
+        DelayMap.unionWith((v1, v2) -> v1 - v2, Map.toDelayMap(Map#{1 -> 2, 2 -> 3}), Map.toDelayMap(Map#{1 -> 8})) |> DelayMap.toMap == Map#{1 -> -6, 2 -> 3}
+
+    @test
+    def unionWith14(): Bool =
+        DelayMap.unionWith((v1, v2) -> v1 - v2, Map.toDelayMap(Map#{1 -> 4}), Map.toDelayMap(Map#{1 -> 2, 2 -> 3})) |> DelayMap.toMap == Map#{1 -> 2, 2 -> 3}
+
+    @test
+    def unionWith15(): Bool =
+        DelayMap.unionWith((v1, v2) -> v1 - v2, Map.toDelayMap(Map#{2 -> 7}), Map.toDelayMap(Map#{1 -> 2, 2 -> 3})) |> DelayMap.toMap == Map#{1 -> 2, 2 -> 4}
+
+    @test
+    def unionWith16(): Bool =
+        DelayMap.unionWith((v1, v2) -> v1 - v2, Map.toDelayMap(Map#{2 -> 7, 11 -> 14, 9 -> 8, 15 -> 22}), Map.toDelayMap(Map#{15 -> 21, 1 -> 2, 2 -> 8, 44 -> 33})) |> DelayMap.toMap == Map#{2 -> -1, 11 -> 14, 9 -> 8, 15 -> 1, 1 -> 2, 44 -> 33}
+
+
+    /////////////////////////////////////////////////////////////////////////////
+    // unionWithKey (pure)                                                     //
+    /////////////////////////////////////////////////////////////////////////////
+
+    @test
+    def unionWithKeyPure01(): Bool =
+        DelayMap.unionWithKey((k, v1, v2) -> k + v1 - v2, Map.toDelayMap(Map#{}), Map.toDelayMap(Map#{})) |> DelayMap.toMap == Map#{}: Map[Int, Int]
```

Comment:
Maybe here we should have some local variables m1 and m2 to make it easier to read.